source/validate_capability.cpp

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Validates OpCapability instruction.
+
+#include "validate.h"
+
+#include <cassert>
+#include <unordered_set>
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+namespace libspirv {
+
+namespace {
+
+bool IsSupportGuaranteedVulkan_1_0(uint32_t capability) {
+  switch (capability) {
+    case SpvCapabilityMatrix:
+    case SpvCapabilityShader:
+    case SpvCapabilityInputAttachment:
+    case SpvCapabilitySampled1D:
+    case SpvCapabilityImage1D:
+    case SpvCapabilitySampledBuffer:
+    case SpvCapabilityImageBuffer:
+    case SpvCapabilityImageQuery:
+    case SpvCapabilityDerivativeControl:
+      return true;
+  }
+  return false;
+}
+
+bool IsSupportOptionalVulkan_1_0(uint32_t capability) {
+  switch (capability) {
+    case SpvCapabilityGeometry:
+    case SpvCapabilityTessellation:
+    case SpvCapabilityFloat64:
+    case SpvCapabilityInt64:
+    case SpvCapabilityInt16:
+    case SpvCapabilityTessellationPointSize:
+    case SpvCapabilityGeometryPointSize:
+    case SpvCapabilityImageGatherExtended:
+    case SpvCapabilityStorageImageMultisample:
+    case SpvCapabilityUniformBufferArrayDynamicIndexing:
+    case SpvCapabilitySampledImageArrayDynamicIndexing:
+    case SpvCapabilityStorageBufferArrayDynamicIndexing:
+    case SpvCapabilityStorageImageArrayDynamicIndexing:
+    case SpvCapabilityClipDistance:
+    case SpvCapabilityCullDistance:
+    case SpvCapabilityImageCubeArray:
+    case SpvCapabilitySampleRateShading:
+    case SpvCapabilitySparseResidency:
+    case SpvCapabilityMinLod:
+    case SpvCapabilitySampledCubeArray:
+    case SpvCapabilityImageMSArray:
+    case SpvCapabilityStorageImageExtendedFormats:
+    case SpvCapabilityInterpolationFunction:
+    case SpvCapabilityStorageImageReadWithoutFormat:
+    case SpvCapabilityStorageImageWriteWithoutFormat:
+    case SpvCapabilityMultiViewport:
+      return true;
+  }
+  return false;
+}
+
+// Checks if |capability| was enabled by extension.
+bool IsEnabledByExtension(ValidationState_t& _, uint32_t capability) {
+  spv_operand_desc operand_desc;
+  const spv_result_t lookup_result = _.grammar().lookupOperand(
+      SPV_OPERAND_TYPE_CAPABILITY, capability, &operand_desc);
+
+  assert(lookup_result == SPV_SUCCESS);
```

Comment:
Either the "if" on the next line is dead, or this assert is bad.  Remove one or the other.

This isn't a public API, so it's guarded by the fact that the binary parser will reject an OpCapability with argument it doesn't recognize.  If in doubt, then a test would exercise it.  You can use !-syntax to exercise it.
E.g. 
`OpCapability !9999999`